### PR TITLE
Unify SymIntNode and SymFloatNode into SymNode

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87817

This refactor was prompted by challenges handling mixed int/float
operations in C++.  A previous version of this patch
added overloads for each permutation of int/float and was unwieldy
https://github.com/pytorch/pytorch/pull/87722/  This PR takes a different
approach.

The general outline of the patch is to combine the C++ types SymIntNode
and SymFloatNode into a single type, SymNode.  This is type erased; we
no longer know statically at C++ if we have an int/float and have to test
it with the is_int()/is_float() virtual methods.  This has a number of
knock on effects.

- We no longer have C++ classes to bind to Python.  Instead, we take an
  entirely new approach to our Python API, where we have a SymInt/SymFloat
  class defined entirely in Python, which hold a SymNode (which corresponds
  to the C++ SymNode).  However, SymNode is not pybind11-bound; instead,
  it lives as-is in Python, and is wrapped into C++ SymNode using PythonSymNode
  when it goes into C++.  This implies a userland rename.

  In principle, it is also possible for the canonical implementation of SymNode
  to be written in C++, and then bound to Python with pybind11 (we have
  this code, although it is commented out.)  However, I did not implement
  this as we currently have no C++ implementations of SymNode.

  Because we do return SymInt/SymFloat from C++ bindings, the C++ binding
  code needs to know how to find these classes.  Currently, this is done
  just by manually importing torch and getting the attributes.

- Because SymInt/SymFloat are easy Python wrappers, __sym_dispatch__ now
  takes SymInt/SymFloat, rather than SymNode, bringing it in line with how
  __torch_dispatch__ works.

Some miscellaneous improvements:

- SymInt now has a constructor that takes SymNode.  Note that this
  constructor is ambiguous if you pass in a subclass of SymNode,
  so an explicit downcast is necessary.  This means toSymFloat/toSymInt
  are no more.  This is a mild optimization as it means rvalue reference
  works automatically.

- We uniformly use the caster for c10::SymInt/SymFloat, rather than
  going the long way via the SymIntNode/SymFloatNode.

- Removed some unnecessary toSymInt/toSymFloat calls in normalize_*
  functions, pretty sure this doesn't do anything.

- guard_int is now a free function, since to guard on an int you cannot
  assume the method exists.  A function can handle both int and SymInt
  inputs.

- We clean up the magic method definition code for SymInt/SymFloat/SymNode.
  ONLY the user classes (SymInt/SymFloat) get magic methods; SymNode gets
  plain methods; this is to help avoid confusion between the two types.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: c10/core/SymInt.cpp [hunk 1]
```diff
@@ -1,95 +1,94 @@
...
-SymIntNode SymInt::toSymIntNodeImpl() const {
+SymNode SymInt::toSymNodeImpl() const {
   TORCH_CHECK(is_symbolic());
-  return SymIntNode::reclaim_copy(toSymIntNodeImplUnowned());
+  return SymNode::reclaim_copy(toSymNodeImplUnowned());
 }
 
-c10::SymInt SymInt::toSymInt(SymIntNode sin_sp) {
+SymInt::SymInt(SymNode sin_sp) {
+  TORCH_CHECK(sin_sp->is_int());
```
Comment (bdhirsh) [inline c10/core/SymInt.cpp:32]: should this be an internal assert? or can user code end up doing this

Comment (ezyang) [reply c10/core/SymInt.cpp:32]: I think this shouldn't be possible to exercise from Python, but you can definitely trigger this relatively easily in c++. No big pref from me.

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -479,8 +478,8 @@ def inner_torch_dispatch(self, func, types, args=(), kwargs=None):
         return out
 
 
-SymInt = torch.SymIntNode
-SymFloat = torch.SymFloatNode
+SymInt = torch.SymInt
```
Comment (albanD) [inline torch/fx/experimental/proxy_tensor.py]: can just be removed now that they are defined above.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -88,56 +89,88 @@ def _handle_sym_dispatch(func, args, kwargs):
...
+        return SymNode(sympy.Integer(num), self.shape_env, float, constant=num)
 
     def clone(self):
-        return PySymInt(self.expr, self.shape_env, constant=self.constant)
+        return SymNode(self.expr, self.shape_env, self.pytype, constant=self.constant)
 
-    def _update_expr(self):
-        self._expr = self.shape_env.replace(self._expr)
+    def str(self):
+        return f"{self.expr}"
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:164]: Do we want a `SymNode()` around? Or we will make nice printing on the SymInt/SymFloat in the future and that will be enough.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -258,19 +298,34 @@ def eval(cls, a):
...
         if SYM_FUNCTION_MODE:
-            return _handle_sym_dispatch(op, (self, other), {})
-        if isinstance(other, py_type):
+            r = _handle_sym_dispatch(op, (wrap_node(self), wrap_node(other)), {})
+            assert isinstance(r, (SymInt, SymFloat)), type(r)
+            return r.node
+        if isinstance(other, SymNode):
             other_expr = other.expr
         else:
+            # TODO: can this case actually occur :think:
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:328]: I don't understand it either. Happy to assert SymNode and remove this.

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -410,9 +489,7 @@ def create_symbolic_sizes_strides(self, ex: torch.Tensor):
         return [self.create_symintnode(i) for i in size], [self.create_symintnode(i) for i in stride]  # type: ignore[arg-type]
 
     def create_symintnode(self, expr: Union["sympy.Expr", int]):
-        py_sym_int = PySymInt(expr, self)
-        cpp_sym_int = torch.SymIntNode.new_symint(py_sym_int)  # type: ignore[attr-defined]
-        return cpp_sym_int
+        return SymInt(SymNode(expr, self, int))
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:492]: Do we prefer `wrap_node(SymNode(expr, self, int))` ?

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -279,45 +334,69 @@ def magic_impl(self, other):
...
+
+    if method in unary_magic_methods:
+        setattr(user_type, f"__{method}__", unary_magic_impl)
+    else:
+        setattr(user_type, f"__{method}__", binary_magic_impl)
         if method in reflectable_magic_methods:
-            setattr(py_type, f"__r{method}__", magic_impl)
+            setattr(user_type, f"__r{method}__", rbinary_magic_impl)
+
+for method, func in magic_methods.items():
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:390]: nit: move this above _make_user_magic so that it is easier to see that they are independent.

**Review by albanD — APPROVED** [review]
Sounds good!

**Comment by albanD** [comment]
PR sanity check can be ignored as this is a refactor.

File: torch/__init__.py [hunk 7]
```diff
@@ -196,6 +196,64 @@ def _load_global_deps():
...
+
+    # for BC
+    def get_pyobj(self):
+        return self.node
+
+class SymFloat:
+    """Like an float, but symbolically records operations on it.  Users interact
+    with this.  Has magic methods."""
+
+    # MUST be named this, C++ binding relies on it
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/__init__.py:233]: i'll update, but essentially it is saying you can't rename "node", it's hard coded in C++ too

File: torch/__init__.py [hunk 8]
```diff
@@ -196,6 +196,64 @@ def _load_global_deps():
 if TYPE_CHECKING:
     import torch._C as _C
 
+class SymInt:
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/__init__.py:199]: yeah, there's some import order shenanigans. There might be another resolution but this ordering works

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:164]: I am ambivalent about this. On the one hand, wrapping with SymNode makes it more obvious with repr that you have a SymNode vs SymInt/SymFloat, and this is useful when debugging type confusion. OTOH, when you actually print (e.g., from C++) you really just want something clean like "2" or "s0 + s1". Let's keep it like this for now (as it preserves old behavior) and see if we want to litigate it later.

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:492]: it doesn't really matter

File: functorch/experimental/cond.py [hunk 9]
```diff
@@ -18,6 +18,8 @@
 
 def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
     def _unwrap_proxy(e):
+        if not isinstance(e, (torch.Tensor, torch.SymInt, torch.SymFloat)):
+            return e
```
Comment (anjali411) [inline functorch/experimental/cond.py:22]: Is this a bug fix? How was it caught?

File: torch/csrc/jit/python/init.cpp [hunk 10]
```diff
@@ -1381,276 +1142,41 @@ void initJITBindings(PyObject* module) {
...
-            }
-            return py::none();
-          })
-      .def("__str__", [](c10::SymFloatNode a) { return a->str(); });
+  // NB: This isn't actually used for regular PyTorch symbolic tracing;
+  // XLA is what needs this
+#define SYMNODE_UNARY(n) .def(#n, [](c10::SymNode a) { return a->n(); })
+#define SYMNODE_UNARY2(n2, n) .def(#n2, [](c10::SymNode a) { return a->n(); })
+#define SYMNODE_BINARY(n) \
+  .def(#n, [](c10::SymNode a, c10::SymNode b) { return a->n(b); })
```
Comment (anjali411) [inline torch/csrc/jit/python/init.cpp:1150]: what about SymNode x PyObject ops?

**Review by anjali411 — APPROVED** [review]

File: functorch/experimental/cond.py (hunk 9 above)
Comment (ezyang) [reply functorch/experimental/cond.py:22]: The unit tests. It's not strictly a bug, but I made the asserts for proxy retrieval more strict and cond was sending us garbage previously.

File: torch/csrc/jit/python/init.cpp (hunk 10 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:1150]: The invariant is that SymNode methods ONLY work with other SymNodes; the top level classes SymInt/SymFloat handle making sure wrapping an int into a SymNode

File: tools/autograd/gen_autograd_functions.py [hunk 11]
```diff
@@ -291,7 +291,7 @@
 for (auto i : c10::irange(prop.size())) {
     auto si = prop[i];
     if (si.is_symbolic()) {
-      auto py_symint = py::cast(si.toSymIntNodeImpl()).release().ptr();
+      auto py_symint = py::cast(si).release().ptr();
```
[the remark this answers is not in the document]
Comment (ezyang) [reply tools/autograd/gen_autograd_functions.py:294]: no, all of those toSymNodeImpl were unnecessary, the new design is much easier.

**Comment by ezyang** [comment]
full green except for XLA: https://hud.pytorch.org/pytorch/pytorch/pull/87817?sha=0a6c135dcd2c05db5e040196feac07cd963abd46

subsequent pushes are xla pin update only
